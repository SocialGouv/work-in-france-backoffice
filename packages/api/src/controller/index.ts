export * from "./alert.controller";
export * from "./monthly-report.controller";
export * from "./validity-check.controller";
