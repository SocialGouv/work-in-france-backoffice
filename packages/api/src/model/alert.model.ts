import { Email } from "../service";
import { IIdentifiable } from "../util";
import { DSGroup } from "./dossier-record.model";

export type AlertType =
  | "closedWithoutDateDebutOrDateFin"
  | "closedWithDebutSupFin"
  | "closedWithSupOneYear"
  | "closedAndMessageReceived"
  | "refusedAndMessageReceived"
  | "withoutContinuationAndMessageReceived"
  | "receivedAndDelayTooLong"
  | "initiatedAndDelayTooLong";

  export type AlertEmailState = "to_send" | "sent" | "blocked";

export interface Alert extends IIdentifiable {
  ds_key: string;
  url: string;
  group: DSGroup;
  alert_type: AlertType;
  message: string;
  instructors_history: string[];
  email_id?: string;
  email_usager: string;
  email_instructors: string[];
  date_debut_apt: string;
  processed_at: number | null;
  email?: Email;
  email_state?: AlertEmailState;
  email_processed_at?: number;
  state: string; // dossier state
}

export interface AlertMessage {
  message: string;
}

export const alertMaxReceivedTimeInDays = 7;
export const alertMaxInitiatedTimeInDays = 7;

export const alertMessages = {
  closedWithoutDateDebutOrDateFin:
    "dossier accepté, date de début et/ou de fin APT manquante",

  closedWithDebutSupFin:
    "dossier accepté, date de fin APT antérieure à Date de début APT",
  closedWithSupOneYear: "dossier accepté, durée APT supérieure à 12 mois",

  closedAndMessageReceived:
    "dossier archivé et message envoyé après acceptation",
  refusedAndMessageReceived: "dossier archivé et message envoyé après refus",
  withoutContinuationAndMessageReceived:
    "dossier archivé et message après classement sans suite",

  receivedAndDelayTooLong: "durée d'instruction de dossier dépassée",

  initiatedAndDelayTooLong: "durée de construction de dossier dépassée"
};
