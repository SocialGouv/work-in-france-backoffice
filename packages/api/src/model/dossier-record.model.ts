import { differenceInDays } from "date-fns";
import { asDate, IIdentifiable, logger } from "../util";

export interface DSGroup {
  id: string;
  label: string;
}

export interface DSChamp {
  value: string;
  type_de_champ: {
    libelle: string;
  };
}

export interface DSCommentaire {
  body: string;
  email: string;
  attachement: string | null;
  created_at: string;
}

interface DSEtablissement {
  siret: string;
}

export interface DSDossier {
  id: number;
  etablissement: DSEtablissement;
  champs: DSChamp[];
  champs_private: DSChamp[];
  commentaires: DSCommentaire[];
  instructeurs: string[];
  email: string;
  archived: boolean;
}

export interface DossierRecord extends IIdentifiable {
  ds_key: string;
  ds_data: DSDossier;
  metadata: {
    state: string;
    procedure_id: string;
    group: DSGroup;
    // date de création du dossier
    created_at: number;
    // date de depot du dossier
    initiated_at: number | null;
    // date de la dernière modification du dossier
    updated_at: number | null;
    // date du passage en instruction
    received_at: number | null;
    // date de décision du dossier
    processed_at: number | null;
    instructors_history: string[];
  };
}

const getFieldValue = (champs: DSChamp[], libelle: string) => {
  const field = champs.find(
    (f: DSChamp) => f.type_de_champ.libelle === libelle
  );
  if (!field) {
    throw new Error(`field ${libelle} does not exists`);
  }
  return field.value;
};

const getPublicFieldValue = (record: DossierRecord, libelle: string) => {
  try {
    return getFieldValue(record.ds_data.champs, libelle);
  } catch (err) {
    logger.error(
      `public field ${libelle} not find for dossier ${record.ds_key}`,
      err
    );
    throw err;
  }
};
const getPrivateFieldValue = (record: DossierRecord, libelle: string) => {
  try {
    return getFieldValue(record.ds_data.champs_private, libelle);
  } catch (err) {
    logger.error(
      `private field ${libelle} not find for dossier ${record.ds_key}`,
      err
    );
    throw err;
  }
};

export const hasExpired = (dossier: DossierRecord): boolean => {
  const endDate = getDateFinAPT(dossier);
  if (!endDate) {
    return true;
  }
  if (endDate < new Date()) {
    return true;
  }
  return false;
};

export const getDateDebutAPTValue = (doc: DossierRecord) =>
  getPrivateFieldValue(doc, "Date de début APT");
export const getDateFinAPTValue = (doc: DossierRecord) =>
  getPrivateFieldValue(doc, "Date de fin APT");
export const getPrenomValue = (doc: DossierRecord) =>
  getPublicFieldValue(doc, "Prénom");
export const getNomValue = (doc: DossierRecord) =>
  getPublicFieldValue(doc, "Nom");
export const getDateNaissanceValue = (doc: DossierRecord) =>
  getPublicFieldValue(doc, "Date de naissance");

export const getDateFinAPT = (dossier: DossierRecord) =>
  asDate(getDateFinAPTValue(dossier));
export const getDateDebutAPT = (dossier: DossierRecord) =>
  asDate(getDateDebutAPTValue(dossier));

export const getDateDebutInstruction = (doc: DossierRecord) =>
  asDate(doc.metadata.received_at);
export const getDateDebutConstruction = (doc: DossierRecord) =>
  asDate(doc.metadata.initiated_at);

export const isLong = (doc: DossierRecord) => {
  const startDate = getDateDebutAPT(doc);
  const endDate = getDateFinAPT(doc);
  if (!startDate || !endDate) {
    return false;
  }
  return differenceInDays(endDate, startDate) + 1 >= 91;
};

const getState = (dossier: DossierRecord) => dossier.metadata.state;

export const isInitiated = (dossier: DossierRecord) =>
  getState(dossier) === "initiated";
export const isReceived = (dossier: DossierRecord) =>
  getState(dossier) === "received";
export const isClosed = (dossier: DossierRecord) =>
  getState(dossier) === "closed";
export const isRefused = (dossier: DossierRecord) =>
  getState(dossier) === "refused";
export const isWithoutContinuation = (dossier: DossierRecord) =>
  getState(dossier) === "without_continuation";

export const getNationality = (dossier: DossierRecord) =>
  getPublicFieldValue(dossier, "Nationalité");

export const getDemarcheSimplifieeUrl = (dossier: DossierRecord) => {
  const keys = dossier.ds_key.split("-");
  const procedureId = keys[0];
  const dossierId = keys[1];
  return `https://www.demarches-simplifiees.fr/procedures/${procedureId}/dossiers/${dossierId}`;
};
