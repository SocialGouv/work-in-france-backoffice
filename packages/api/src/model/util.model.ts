export const getDossierId = (dsKey: string) => {
  return dsKey.split("-")[1];
};
