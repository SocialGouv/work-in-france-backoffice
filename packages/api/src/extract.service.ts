import { addMonths, getMonth, getYear, isBefore } from "date-fns";
import { Observable, Observer, Subject } from "rxjs";
import { concatMap, exhaustMap } from "rxjs/operators";
import { Alert } from "./model";
import { MonthlyReport } from "./model/monthly-report.model";
import { addAlerts } from "./scheduler/alert.scheduler";
import { monthlyreportService } from "./service";
import { logger } from "./util";
import { YearMonth } from "./util/interface/year-month";

class ExtractorService {
  private syncAllMonthlyReports$ = new Subject();
  private syncAllAlerts$ = new Subject();

  constructor() {
    this.initAllMonthlyReportSynchro();
    this.initAllAlertSynchro();
  }

  public launchGlobalMonthlyReportSynchro() {
    this.syncAllMonthlyReports$.next();
  }

  public launchGlobalAlertSynchro() {
    this.syncAllAlerts$.next();
  }

  private allMonthlyReports(): Observable<MonthlyReport> {
    return allDates$.pipe(
      concatMap((res: YearMonth) =>
        monthlyreportService.syncMonthlyReports(res.year, res.month)
      )
    );
  }

  private initAllAlertSynchro() {
    this.syncAllAlerts$
      .pipe(concatMap(() => addAlerts(new Date(0))))
      .subscribe({
        error: (error: Error) => logger.error(`[alerts synchro] error`, error),
        next: (alert: Alert) =>
          logger.info(
            `[alerts synchro] alert added ${alert.ds_key} ${alert.message}`
          )
      });
  }

  private initAllMonthlyReportSynchro() {
    this.syncAllMonthlyReports$
      .pipe(exhaustMap(() => this.allMonthlyReports()))
      .subscribe({
        complete: () => logger.info(`[monthly reports synchro] completed`),
        error: (err: any) =>
          logger.error(`[monthly reports synchro] error: `, err),
        next: (next: MonthlyReport) =>
          logger.info(
            `[monthly reports synchro] report ${next.year}-${next.month} ${next.group.label} synchronised `
          )
      });
  }
}

const allDates$ = Observable.create((observer: Observer<YearMonth>) => {
  let start = new Date(2018, 1, 1);
  const onMonthBefore = addMonths(new Date(), -1);
  while (isBefore(start, onMonthBefore)) {
    observer.next({ year: getYear(start), month: getMonth(start) });
    start = addMonths(start, 1);
  }
  observer.complete();
});

export const extractorService = new ExtractorService();
