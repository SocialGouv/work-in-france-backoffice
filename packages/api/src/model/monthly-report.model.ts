import { IIdentifiable } from "../util";
import { DSGroup } from "./dossier-record.model";

export interface MonthlyReportCounter {
  count: number;
  countries: {
    [country: string]: number;
  };
  dossiers: string[];
}

export interface MonthlyReport extends IIdentifiable {
  year: number;
  month: number;
  group: DSGroup;
  accepted: {
    more3Months: MonthlyReportCounter;
    less3Months: MonthlyReportCounter;
  };
  refused: MonthlyReportCounter;
  withoutContinuation: MonthlyReportCounter;
}

const initMonthlyReportCounter: () => MonthlyReportCounter = () => ({
  count: 0,
  countries: {},
  dossiers: [],
});

export const initReport: (
  year: number,
  month: number,
  group: DSGroup
) => MonthlyReport = (year: number, month: number, group: DSGroup) => {
  return {
    year,
    // tslint:disable-next-line: object-literal-sort-keys
    month,
    group,
    accepted: {
      less3Months: initMonthlyReportCounter(),
      more3Months: initMonthlyReportCounter(),
    },
    refused: initMonthlyReportCounter(),
    withoutContinuation: initMonthlyReportCounter(),
  };
};
