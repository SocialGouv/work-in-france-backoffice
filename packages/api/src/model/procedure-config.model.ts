import { DSType } from "./dossier-record.model";

export interface ProcedureConfig {
  procedures: number[];
  group: {
    id: string;
    label: string;
    type: DSType;
  };
}
