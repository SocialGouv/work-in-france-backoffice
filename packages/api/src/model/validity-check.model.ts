import { IIdentifiable } from "../util/persistence";
import { DSChamp, DSType } from "./dossier-record.model";

export interface ValidityCheck extends IIdentifiable {
  type: DSType;
  dossier_id: number;
  ds_key: string;
  siret: string;
  prenom: string;
  nom: string;
  date_de_naissance: string;
  has_expired: boolean;
  date_de_debut_apt: string;
  date_de_fin_apt: string;
  fin_apt: Date;
  champs: DSChamp[];
}

// "ds_id": 378335,
// "siret": "000000000000000",
// "prenom": "*y***",
// "nom": "*a******",
// "date_de_naissance": "1990-01-01",
// "has_expired": false,
// "date_de_debut_apt": "2019-03-18",
// "date_de_fin_apt": "2019-11-04"
