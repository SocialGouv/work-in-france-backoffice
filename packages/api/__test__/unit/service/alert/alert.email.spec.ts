import { Alert } from "../../../../src/model";
import { getAlertEmail } from "../../../../src/service";

const alert: Alert = {
  url: "URL",
  // tslint:disable-next-line: object-literal-sort-keys
  group: {
    id: "75",
    label: "75 - Paris"
  },
  ds_key: "14131-502017",
  alert_type: "closedWithDebutSupFin",
  email_usager: "usager@wif.fr",
  date_debut_apt: "2019-01-01",
  email_instructors: [
    "instructor2@direccte.gouv.fr",
    "instructor3@direccte.gouv.fr"
  ],
  instructors_history: [
    "instructor1@direccte.gouv.fr",
    "instructor2@direccte.gouv.fr",
    "instructor3@direccte.gouv.fr"
  ],
  message: "ALERT_MESSAGE",
  processed_at: 0,
  state: "initiated"
};

it("email of alert should be equal to", () => {
  expect(
    getAlertEmail({ ...alert, alert_type: "closedWithDebutSupFin" })
  ).toMatchSnapshot();

  expect(
    getAlertEmail({ ...alert, alert_type: "closedWithoutDateDebutOrDateFin" })
  ).toMatchSnapshot();

  expect(
    getAlertEmail({ ...alert, alert_type: "closedWithSupOneYear" })
  ).toMatchSnapshot();

  expect(
    getAlertEmail({ ...alert, alert_type: "initiatedAndDelayTooLong" })
  ).toBeUndefined();

  expect(
    getAlertEmail({ ...alert, alert_type: "receivedAndDelayTooLong" })
  ).toBeUndefined();

  expect(
    getAlertEmail({ ...alert, alert_type: "closedAndMessageReceived" })
  ).toBeUndefined();

  expect(
    getAlertEmail({ ...alert, alert_type: "refusedAndMessageReceived" })
  ).toBeUndefined();

  expect(
    getAlertEmail({
      ...alert,
      alert_type: "withoutContinuationAndMessageReceived"
    })
  ).toBeUndefined();
});
