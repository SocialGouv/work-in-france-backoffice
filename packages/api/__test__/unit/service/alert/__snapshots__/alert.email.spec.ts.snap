// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email of alert should be equal to 1`] = `
Object {
  "bcc": Array [
    Object {
      "email": "contact@workinfrance.beta.gouv.fr",
      "name": "contact@workinfrance.beta.gouv.fr",
    },
  ],
  "bodyText": "Bonjour,

Nous avons identifié une erreur sur votre autorisation provisoire de travail n°502017. En effet, la date de fin est antérieure à la date de début de votre autorisation qui, par conséquent, n'est pas valide.

Selon l'article L.242-1 du Code des relations entre le public et l'administration (https://www.legifrance.gouv.fr/affichCodeArticle.do?cidTexte=LEGITEXT000031366350&idArticle=LEGIARTI000031367657&dateTexte=&categorieLien=cid) l'administration ne peut revenir sur une décision créatrice de droits. Ainsi, les agents du service de la main d'oeuvre étrangère de la Direccte ne peuvent revenir sur cette décision.

Nous vous suggérons de déposer un nouveau dossier le plus rapidement possible.

Bonne journée,


L'équipe WorkInFrance


—


Merci de ne pas répondre à cet email",
  "cci": Array [
    Object {
      "email": "instructor2@direccte.gouv.fr",
      "name": "instructor2@direccte.gouv.fr",
    },
    Object {
      "email": "instructor3@direccte.gouv.fr",
      "name": "instructor3@direccte.gouv.fr",
    },
  ],
  "subject": "Erreur sur votre APT WorkInFrance nº502017 / UD075",
  "to": Array [
    Object {
      "email": "usager@wif.fr",
      "name": "usager@wif.fr",
    },
  ],
}
`;

exports[`email of alert should be equal to 2`] = `
Object {
  "bcc": Array [
    Object {
      "email": "contact@workinfrance.beta.gouv.fr",
      "name": "contact@workinfrance.beta.gouv.fr",
    },
  ],
  "bodyText": "Bonjour,

Nous avons identifié une erreur sur votre autorisation provisoire de travail nº502017. En effet, il n'y a pas de date de début et/ou de date de fin sur votre autorisation qui, par conséquent, n'est pas valide.

Selon l'article L.242-1 du Code des relations entre le public et l'administration (https://www.legifrance.gouv.fr/affichCodeArticle.do?cidTexte=LEGITEXT000031366350&idArticle=LEGIARTI000031367657&dateTexte=&categorieLien=cid) l'administration ne peut revenir sur une décision créatrice de droits. Ainsi, les agents du service de la main d'oeuvre étrangère de la Direccte ne peuvent revenir sur cette décision.

Nous vous suggérons de déposer un nouveau dossier le plus rapidement possible.

Bonne journée,


L'équipe WorkInFrance


—


Merci de ne pas répondre à cet email",
  "cci": Array [
    Object {
      "email": "instructor2@direccte.gouv.fr",
      "name": "instructor2@direccte.gouv.fr",
    },
    Object {
      "email": "instructor3@direccte.gouv.fr",
      "name": "instructor3@direccte.gouv.fr",
    },
  ],
  "subject": "Erreur sur votre APT WorkInFrance nº502017 / UD075",
  "to": Array [
    Object {
      "email": "usager@wif.fr",
      "name": "usager@wif.fr",
    },
  ],
}
`;

exports[`email of alert should be equal to 3`] = `
Object {
  "bcc": Array [
    Object {
      "email": "contact@workinfrance.beta.gouv.fr",
      "name": "contact@workinfrance.beta.gouv.fr",
    },
  ],
  "bodyText": "Bonjour,

Nous avons identifié une erreur sur votre autorisation provisoire de travail n°502017. En effet, la durée de votre autorisation est supérieure à 12 mois.

Selon l'alinéa 14 de l'article R.5221-3 du Code du travail (https://www.legifrance.gouv.fr/affichCodeArticle.do?idArticle=LEGIARTI000038189414&cidTexte=LEGITEXT000006072050&dateTexte=20190301) une autorisation provisoire de travail a une durée maximale de douze mois.

Nous vous suggérons de déposer un nouveau dossier à partir du 27/11/2019

Bonne journée,


L'équipe WorkInFrance


—


Merci de ne pas répondre à cet email",
  "cci": Array [
    Object {
      "email": "instructor2@direccte.gouv.fr",
      "name": "instructor2@direccte.gouv.fr",
    },
    Object {
      "email": "instructor3@direccte.gouv.fr",
      "name": "instructor3@direccte.gouv.fr",
    },
  ],
  "subject": "Erreur sur votre APT WorkInFrance nº502017 / UD075",
  "to": Array [
    Object {
      "email": "usager@wif.fr",
      "name": "usager@wif.fr",
    },
  ],
}
`;
